### PR TITLE
add admin view of new tables

### PR DESCRIPTION
**Pull request by diitaz93**
## Description

Closes https://github.com/Clinical-Genomics/add-new-tech/issues/22

Add table **Illumina Flow Cell** to StatusDB with the same columns as the **Flowcell** table and the same operations (search and sort).

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change) or deploy locally with script

### How to test

- [x] Do deploy in the VM or locally

### Expected test outcome

- [x] Check that the table is rendered correctly, the search and sort operations work

## Review

- [x] Tests executed by SD
- [x] "Merge and deploy" approved by CO
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **MINOR** - when you add functionality in a backwards compatible manner

## Implementation Plan

- [ ] Deployed to stage:
```shell

```
- [ ] Deployed to production:
```shell

```
- [x] Inform to @Clinical-Genomics/prodbioinfo

### CODE REVIEW

**Comment by diitaz93** [comment]
## Test on cg-API stage
Screenshot 2024-05-23 at 16 03 55

File: cg/store/models.py [hunk 1]
```diff
@@ -1051,6 +1051,19 @@ class IlluminaSequencingRun(InstrumentRun):
 
     __mapper_args__ = {"polymorphic_identity": DeviceType.ILLUMINA}
 
+    @hybrid_property
```
Comment (seallard) [inline cg/store/models.py:1054]: What are these decorators? Never saw them before

File: cg/store/models.py [hunk 2]
```diff
@@ -1051,6 +1051,19 @@ class IlluminaSequencingRun(InstrumentRun):
...
+    def flow_cell(self):
+        return self.device.internal_id if self.device else None
+
+    @flow_cell.expression
```
Comment (seallard) [inline cg/store/models.py:1058]: How does this decorator work?

File: cg/store/models.py (hunk 1 above)
Comment (diitaz93) [reply cg/store/models.py:1054]: `flow_cell` is not an attribute of the model, so this decorator allows SQLAlchemy to recognize `flow_cell` as an attribute

File: cg/server/admin.py [hunk 3]
```diff
@@ -441,6 +442,49 @@ class AnalysisView(BaseView):
     form_extra_fields = {"workflow": SelectEnumField(enum_class=Workflow)}
 
 
+class IlluminaFlowCellView(BaseView):
```
Comment (ChrOertlin) [inline cg/server/admin.py:445]:
```suggestion
class IlluminaSequencingRunView(BaseView):
```

Comment (ChrOertlin) [reply cg/server/admin.py:445]: The flowcell is the run device model.

File: cg/server/admin.py [hunk 4]
```diff
@@ -441,6 +442,49 @@ class AnalysisView(BaseView):
...
+    """Admin view for Model.IlluminaSequencingRun"""
+
+    column_list = (
```
Comment (ChrOertlin) [inline cg/server/admin.py:448]: I think we should extend these to also show the aggregated metrics i.e. reads/ yield for the run so production can easily see get these

File: cg/server/app.py [hunk 5]
```diff
@@ -128,6 +128,9 @@ def _register_admin_views():
     ext.admin.add_view(admin.FlowcellView(Flowcell, ext.db.session))
     ext.admin.add_view(admin.AnalysisView(Analysis, ext.db.session))
     ext.admin.add_view(admin.InvoiceView(Invoice, ext.db.session))
+    ext.admin.add_view(
+        admin.IlluminaFlowCellView(IlluminaSequencingRun, ext.db.session, name="Illumina Flow Cell")
+    )
```
Comment (ChrOertlin) [inline cg/server/app.py:133]: I am confused... the FlowCell is the RunDevice now, is it not?

File: cg/server/admin.py (hunk 3 above)
Comment (diitaz93) [reply cg/server/admin.py:445]: The idea is that the view in the web API the table is called FlowCell, not Sequencing metrics

File: cg/store/models.py (hunk 1 above)
Comment (diitaz93) [reply cg/store/models.py]: replaced for a simpler solution

File: cg/store/models.py (hunk 2 above)
Comment (diitaz93) [reply cg/store/models.py]: replaced for a simpler solution

File: cg/server/admin.py (hunk 4 above)
Comment (diitaz93) [reply cg/server/admin.py:453]: add all columns

File: cg/server/admin.py (hunk 3 above)
Comment (diitaz93) [reply cg/server/admin.py:450]: resolved in person

File: cg/server/app.py (hunk 5 above)
Comment (diitaz93) [reply cg/server/app.py:133]: resolved in person

**Comment by diitaz93** [comment]
## Updated tests
Search works for internal id, sortable columns, filters ✅ 
Screenshot 2024-05-24 at 10 51 28

**Review by ChrOertlin — APPROVED** [review]
Looks good to me!

**Review by Vince-janv — APPROVED** [review]
Much clearer 👍

File: cg/server/admin.py [hunk 6]
```diff
@@ -441,6 +447,53 @@ class AnalysisView(BaseView):
...
+        "internal_id",
+        "sequencer_type",
+        "sequencer_name",
+        "sequenced_at",
```
Comment (ChrOertlin) [inline cg/server/admin.py:457]:
```suggestion
```

File: cg/server/admin.py [hunk 7]
```diff
@@ -441,6 +447,53 @@ class AnalysisView(BaseView):
...
+        "data_availability",
+        "has_backup",
+        "total_reads",
+        "total_undetermined_reads",
```
Comment (ChrOertlin) [inline cg/server/admin.py:461]:
```suggestion
        "total_undetermined_reads",
        "percent_undetermined_reads",
```

File: cg/server/admin.py [hunk 8]
```diff
@@ -441,6 +447,53 @@ class AnalysisView(BaseView):
...
+    column_formatters = {
+        "internal_id": view_flow_cell_internal_id,
+    }
+
+    column_default_sort = ("sequenced_at", True)
+    column_filters = ["sequencer_type", "sequencer_name", "data_availability"]
+    column_editable_list = ["data_availability"]
+    column_searchable_list = ["sequencer_type", "sequencer_name", "device.internal_id"]
+    column_sortable_list = [
+        "sequenced_at",
```
Comment (ChrOertlin) [inline cg/server/admin.py:484]:
```suggestion
```

File: cg/server/admin.py (hunk 6 above)
Comment (diitaz93) [reply cg/server/admin.py:457]: fixed

File: cg/server/admin.py (hunk 7 above)
Comment (diitaz93) [reply cg/server/admin.py:460]: fixed

File: cg/server/admin.py [hunk 9]
```diff
@@ -33,6 +33,12 @@ def view_priority(unused1, unused2, model, unused3):
     return Markup("%s" % model.priority.name) if model else ""
 
 
+def view_flow_cell_internal_id(unused1, unused2, model, unused3):
+    """column formatter for priority"""
+    del unused1, unused2, unused3
+    return Markup("%s" % model.device.internal_id)
```
[the remark this answers is not in the document]
Comment (diitaz93) [reply cg/server/admin.py:39]: not sure, but I consider this out of scope, this function is directly copied from the `Flowcell` model

[the remark this answers is not in the document]
Comment (seallard) [reply cg/server/admin.py:39]:
```python
return Markup(f"{model.device.internal_id}")
```
is the exact same thing

**Pull request merged**